wgpu-hal/src/gles/egl.rs

```diff
@@ -225,6 +225,77 @@ enum SrgbFrameBufferKind {
     Khr,
 }
 
+pub struct AdapterContext {
+    glow_context: glow::Context,
+    egl: Arc<egl::DynamicInstance<egl::EGL1_4>>,
+    egl_display: egl::Display,
+    egl_context: egl::Context,
+    egl_pbuffer: Option<egl::Surface>,
+    mutex: Mutex<()>,
+}
+
+unsafe impl Sync for AdapterContext {}
+unsafe impl Send for AdapterContext {}
+
+pub struct AdapterContextLock<'a> {
+    glow_context: &'a glow::Context,
+    egl: &'a Arc<egl::DynamicInstance<egl::EGL1_4>>,
+    egl_display: egl::Display,
+    egl_pbuffer: Option<egl::Surface>,
+    _mutex_guard: MutexGuard<'a, ()>,
+}
+
+impl<'a> std::ops::Deref for AdapterContextLock<'a> {
+    type Target = glow::Context;
+
+    fn deref(&self) -> &Self::Target {
+        &self.glow_context
+    }
+}
+
+impl<'a> Drop for AdapterContextLock<'a> {
+    fn drop(&mut self) {
+        // Make the EGL context *not* current on this thread
+        self.egl
+            .make_current(self.egl_display, self.egl_pbuffer, self.egl_pbuffer, None)
+            .expect("Cannot make EGL context ");
+    }
+}
+
+impl AdapterContext {
+    pub fn get_without_lock<'a>(&'a self) -> &'a glow::Context {
```

Comment:
as we don't want this to be used widely even within the backend, let's make it `unsafe`